### PR TITLE
Add network policies for the Coherence operator

### PR DESCRIPTION
**Pull request by bobdonat**
# Description

This pull request adds the network policies needed for the Coherence operator.
The network policies required are:
1. ingress from Kubernetes API server
2. egress to Kubernetes API server
3. egress to Coherence pods in application namespace

Fixes VZ-2479

# Checklist 

As the author of this PR, I have:

- [x] Checked that I included or updated copyright and license notices in all files that I altered
- [x] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: application-operator/constants/constants.go [hunk 1]
```diff
@@ -54,6 +54,9 @@ const LabelVerrazzanoManagedDefault = "true"
 // LabelIstioInjection - constant for a Kubernetes label that is applied by Verrazzano
 const LabelIstioInjection = "istio-injection"
 
+// LabelVerrazzanoIONamespace - constant for a Kubernetes label that is used by network policies
+const LabelVerrazzanoIONamespace = "verrazzano.io/namespace"
```
Comment (pfmackin) [inline application-operator/constants/constants.go:58]: looks strange with the IO can we just call it LabelVerrazzanoNamespace

File: application-operator/controllers/cohworkload/coherenceworkload_controller.go [hunk 2]
```diff
@@ -579,6 +588,83 @@ func (r *Reconciler) mutateDestinationRule(destinationRule *istioclient.Destinat
...
+		err := r.Update(ctx, appNamespace)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Create a network policy in the verrazzano-system namespace, if it does not already exist.
+	// This network policy is an egress for the Coherence operator to Coherence pods in application namespaces.
+	networkPolicy := &netv1.NetworkPolicy{}
+	appNamespaceName := fmt.Sprintf("%s-%s", appNamespace.Name, appName)
```
Comment (pfmackin) [inline application-operator/controllers/cohworkload/coherenceworkload_controller.go:614]: Might be good to prefix with coh- or else how would we know this is a coherence related network policy

File: application-operator/controllers/cohworkload/coherenceworkload_controller.go [hunk 3]
```diff
@@ -579,6 +588,83 @@ func (r *Reconciler) mutateDestinationRule(destinationRule *istioclient.Destinat
...
+	err := r.Get(ctx, client.ObjectKey{Namespace: constants.VerrazzanoSystemNamespace, Name: appNamespaceName}, networkPolicy)
+	if err != nil && k8serrors.IsNotFound(err) {
```
Comment (pfmackin) [inline application-operator/controllers/cohworkload/coherenceworkload_controller.go:616]: when does this policy get deleted?

File: application-operator/constants/constants.go (hunk 1 above)
Comment (bobdonat) [reply application-operator/constants/constants.go:58]: fixed

File: application-operator/controllers/cohworkload/coherenceworkload_controller.go (hunk 2 above)
Comment (bobdonat) [reply application-operator/controllers/cohworkload/coherenceworkload_controller.go:614]: fixed.

File: application-operator/controllers/cohworkload/coherenceworkload_controller.go (hunk 3 above)
Comment (bobdonat) [reply application-operator/controllers/cohworkload/coherenceworkload_controller.go:616]: I will enter a follow on Jira for this.  There is no application specific owner reference that I am aware which can be set to have the resource automatically cleaned up.  Maybe we can do this in the appconfig webhook.

**Review by pfmackin — APPROVED** [review]

**Pull request merged**